### PR TITLE
feat(RHTAPREL-838): modify create-pyxis-image to support multi-arch images

### PR DESCRIPTION
**Pull request by theflockers**
This PR adds the necessary changes to allow multi-arch images

Signed-off-by: Leandro Mendes <lmendes@redhat.com>

### CODE REVIEW

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (johnbieren) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: You should be able to call just `get-image-architectures` https://github.com/redhat-appstudio/release-service-utils/blob/main/Dockerfile#L48

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: I find the var name confusing. This is not the arch, it's the digest of the image. So I would suggest just `DIGEST` or `IMAGE_DIGEST`.

But more importantly, why do we need need to first get the arch image digest and then do another `skopeo inspect`? You already get the same response with the first one.

So I think you can just do `skopeo inspect --override-os linux --override-arch "$ARCH" "$PULLSPEC" > /tmp/skopeo-inspect-${ARCH}.json` and be done.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: You call the output file a json, but it's not a json. This is the output of the `create_container_image` script. I'm not sure exactly what's in there, but definitely some logging, including the `The image id is` line which we need to parse to get the image id.

And then the resulting pyxis.json should contain a mapping like:
```json
{
    "components": [
        {
            "imageUrl": "image-url",
            "pyxisImages": [
                {
                    "arch": "amd64",
                    "imageID": "abcd1234"
                },
                {
                    "arch": "s390x",
                    "imageID": "abcd1236"
                }
            ]
        }
    ]
}
```

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Yeah. I was wondering that as well, having  it  for each component. In this initial test I only considered a single one just to provoke these good reviews.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: True. I will fix it.

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: So, just looking it again now - correct me if I am wrong - the thing is that the image sha is different for each arch.
I noticed that by using the `--override-arch` it will override it but the sha will the still the same one, so I assumed that by doing it without getting the right sha, it would be enforcing the arch in the wrong image. 

I need more tests anyway to understand properly this procedure.

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: One quick example:

If you inspect `registry.access.redhat.com/ubi9-minimal@sha256:582e18f13291d7c686ec4e6e92d20b24c62ae0fc72767c46f30a69b1a6198055`

```
RAW_OUTPUT='{
    "manifests": [
        {
            "digest": "sha256:06d06f15f7b641a78f2512c8817cbecaa1bf549488e273f5ac27ff1654ed33f0",
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            },
            "size": 429
        },
        {
            "digest": "sha256:199d5daca3dba0a7deaf0086331917dee256089e94272bef5613517d0007f6f5",
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "platform": {
                "architecture": "arm64",
                "os": "linux"
            },
            "size": 429
        },
```

So the digests are different, so I think the proper is getting it for each arch, then enforcing it per each. What maybe we can change to not need to call it again, is adding the digest to the result for `get-image-architectures` - maybe a parameter `--digest` so we don't need to call it multiple times.

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: I think I have a wrong assumption here. The component will be built on top of a single arch, so the image will have a single arch, not a list. I tested with an image that might not reflect the real use case. Trying one of our release-service builds, it has a single arch. So probably I don´t need any loop of archs.

Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: You're mixing two things. The snapshot will have the digests for the top level thing - either the image itself, or the manifest list digest in case of multiarch images. In both cases using that digest is enough to get the specific arch image you want with `--override-arch` and ``--override-os` - you don't need to worry about the digest of the arch image itself.

Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: > I think I have a wrong assumption here. The component will be built on top of a single arch, so the image will have a single arch, not a list. I tested with an image that might not reflect the real use case. Trying one of our release-service builds, it has a single arch. So probably I don´t need any loop of archs.

The whole point of this PR is to start supporting multiarch images, so I don't understand what you mean by saying that you looked at what we build today and you see single arch images. The thing is that we need to support both single arch and multiarch. But from some discussions in the chat, it seems that in the future, our build system will always produce manifest list images even for single arch.

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: I am confused about what we will "receive" from build-service. For instance  in case `A`

- `component A` is built on top of a `amd64` image and `snapshot-A` is created; the release pipeline runs until create-pyxis-image
- `component A` `snapshot-A` is handled by `create-pyxis-image` 

the question comes here:
will `component A` image  contain only `amd64` information or eventually have multiple archs? even though the component was not built for other archs...

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: btw, found the feature architecture paper. I will take a good read at it so I don't need to flood here anymore.

Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: I don't understand the question. Why would we even care how the build system determines what architectures to build? We only care about the result and then we detect what architectures the image has.

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Unf. @mmalina I am not that smart. Sometimes I need to search/read a lot to understand simple things to reach the point you reach just by looking.

Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Don't worry, I search all the time too 🙂

**Comment by theflockers** [comment]
e2e failed on pyxis task. I am running the pipeline locally to fix the issue.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (johnbieren) [inline tasks/create-pyxis-image/create-pyxis-image.yaml:1]: Not sure if this is ready for review or still being worked through, but need new version and README changelog entry

File: tasks/create-pyxis-image/tests/test-create-pyxis-image-one-containerimage-multi-arch.yaml
Comment (johnbieren) [inline tasks/create-pyxis-image/tests/test-create-pyxis-image-one-containerimage-multi-arch.yaml]: looks like it is expected 3 times

File: tasks/create-pyxis-image/tests/test-create-pyxis-image-multi-containerimages-multi-arch.yaml
Comment (johnbieren) [inline tasks/create-pyxis-image/tests/test-create-pyxis-image-multi-containerimages-multi-arch.yaml]: archs or architectures (missing s :slightly_smiling_face: )

**Review by johnbieren — COMMENTED** [review]
Martin knows better than me here, but just some comments

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:1]: It is indeed  "ready", besides the fact I totally forgot to update the README :grimacing:. On it!

Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml:1]: done!

File: tasks/create-pyxis-image/tests/test-create-pyxis-image-one-containerimage-multi-arch.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/tests/test-create-pyxis-image-one-containerimage-multi-arch.yaml]: fixed.

File: tasks/create-pyxis-image/tests/test-create-pyxis-image-multi-containerimages-multi-arch.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/tests/test-create-pyxis-image-multi-containerimages-multi-arch.yaml]: fixed

**Comment by johnbieren** [comment]
LGTM  but I think best to wait for Martin approval (and needs to pass CI of course)

**Comment by theflockers** [comment]
Moving quickly to draft as I will add tests for the json file.

**Comment by mmalina** [comment]
> LGTM but I think best to wait for Martin approval (and needs to pass CI of course)

I'm waiting for Leandro's update on the open threads.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: The variable name was updated and also changed the way I was calling skopeo.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: done

**Comment by theflockers** [comment]
> > LGTM but I think best to wait for Martin approval (and needs to pass CI of course)
> 
> I'm waiting for Leandro's update on the open threads.

OK. threads updated. Please take  look when you can.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: I wonder if we could streamline the PULLSPEC variable here. We first define it by adding "docker://" as a prefix. And now we use it twice with "docker://" and once without - by removing it again.

Maybe it would be nicer to have the variable without "docker://" and add it on the two occasions where it's needed, rather then removing it in this case? WDYT?

So `PULLSPEC="${REPOSITORY}@${DIGEST}"` and then just add the prefix where needed.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: I would suggest using jq args here - that way you can avoid all the escaping (`\"`).

Like: `jq --argjson id $i --arg image ${CONTAINER_IMAGE} '.components[$id] += {"containerImage": "$image", ...}'`

(The reason for the first `--argjson` instead of `--arg` is because otherwise it would be a string and we want it as integer.)

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: Same as above - I would suggest using `--arg` and `--argjson` so that it can be nicer - without `\"`.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: I would probably suggest to have the awk on a separate line and assign a variable and use that with `--arg`. 

Also, that way you don't have to parse it again later.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Thanks, I added more comments, but this is ok.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: This is ok now.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [inline tasks/create-pyxis-image/create-pyxis-image.yaml]: `$JSON_OUTPUT` was already produced by `jq`, so no need to process it with it again. Instead, you can just do:
`tee $(workspaces.data.path)/pyxis.json <<< "$JSON_OUTPUT"` or `echo "$JSON_OUTPUT" | tee $(workspaces.data.path)/pyxis.json`

**Comment by theflockers** [comment]
moving back to draft to work on the changes.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: I agree, so I removed the "docker://"  from the PULLSPEC declaration.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: true. Totally forgot about those `args`  parameters. changed.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: changed.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: makes sense. Added the `IMAGEID` var above `JSON_OUTPUT` and changed to use the `args` as well.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (theflockers) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: changed.

**Comment by theflockers** [comment]
PTAL @mmalina. Thanks.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Cool, thanks.

File: tasks/create-pyxis-image/create-pyxis-image.yaml
Comment (mmalina) [reply tasks/create-pyxis-image/create-pyxis-image.yaml]: Cool.

**Review by mmalina — APPROVED** [review]
It looks good!

**Review by johnbieren — APPROVED** [review]

**Pull request merged**
